az://303 vclock enhancements - *DO NOT MERGE YET*

jonmeredith opened the pull request:
Hey all,

Here's a pull request for node-based vclocks.  It changes the put FSM to work the way it is described in the Dynamo paper and is resistant to vclock explosion (but not sibling explosion) in the face of poorly behaved clients / multiple updates on a partitioned cluster by a client that can see both sides / syntactic put merge.

The changes
- Require a member of the preflist to coordinate a put.
- Complete the put on the local vnode first using a modified vnode put that ensures a conflicting vclock with other nodes in the cluster (which must be true as only the local node can update that vclock).
- Issue the object put successfully put on the local vnode to the remote vnodes.
- Remove syntactic put merge.
- Add an eventual consistency test that verifies this method works correctly in the face of updates during continuous partitioning of the cluster.
- Update the existing put FSM test for new behavior.

Things still to do for put FSM:
- Add a way to generate a shorter vclock nodeid (maybe initially generate from an ethernet MAC then store/retrieve from the ring, need to handle multiple instances per physical node).
- May need to use a modified vclock nodeid for fallback nodes to ensure uniqueness.

Things to do to riak_kv before we can merge
- Make any long-running vnode operation run on another thread.  Operations are serialized on the local vnode, so (for N=3) it is no longer best two latencies out of three, it's latency of local vnode put + best one of two remote vnode put latencies.

For testing it requires riak_core az303-vclock-enhancements checked out for a minor tweak to riak_core_vnode_master:command.  All k/v eunit tests should pass.

Once those changes have been made we can merge this pull request into mainline, but not before.  I'm submitting it for review now so the logic can be checked.

Cheers, Jon

P.S. DO NOT MERGE THIS YET.

Commit:
Eventual consistency test.

Currently just a framework for writing tests modelling clients and
processes.

Commit:
Added get to ec_eqc.

Commit:
Added skeleton put request to ec_eqc.

Commit:
Added priorities to messages - now scrambles delivery order.

Commit:
Refactored request into a record, removed pri from client.

Commit:
Fixed comments.

Commit:
Changed fsm model in ec_eqc to use get/put core.

Commit:
Ensured last request is a get.

Commit:
Check expected values in results.

Found first counterexample.

Failed!
{[0,0,1,0],
 [{client,undefined,[update,get],undefined},
  {client,undefined,[update],undefined}],
 {params,undefined,0,1,1,undefined}}
Params:
{params,undefined,1,1,1,1}
Clients:
[{client,1,
         [{req,0,undefined,{get,[{kv_vnode,1,1}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1}]}},
          {req,1,undefined,{get,[{kv_vnode,1,1}]}}],
         undefined},
 {client,2,
         [{req,0,undefined,{get,[{kv_vnode,1,1}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1}]}}],
         undefined},
 {client,1000001,[{req,1000002,undefined,{get,[{kv_vnode,1,1}]}}],undefined}]
History:
[{deliver_msg,{msg,0,{req,2},{get_fsm,2},{get,[{kv_vnode,1,1}]}}},
 {deliver_msg,{msg,0,{get_fsm,2},{kv_vnode,1,1},{get,2}}},
 {deliver_msg,{msg,0,{kv_vnode,1,1},{get_fsm,2},{r,{error,notfound},1,2}}},
 {deliver_msg,{msg,1,{get_fsm,2},{req,2},{error,notfound}}},
 {result,2,{req,0,2,{get,[{kv_vnode,1,1}]}},{error,notfound}},
 {deliver_msg,
     {msg,0,
         {req,3},
         {put_fsm,3},
         {put,
             [{kv_vnode,1,1}],
             {r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,3>>}],
                 [{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {deliver_msg,
     {msg,0,
         {put_fsm,3},
         {kv_vnode,1,1},
         {put,
             {r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,3>>}],
                 [{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined},
             3,3}}},
 {deliver_msg,{msg,0,{kv_vnode,1,1},{put_fsm,3},{w,1,3}}},
 {deliver_msg,{msg,1,{req,1},{get_fsm,1},{get,[{kv_vnode,1,1}]}}},
 {deliver_msg,{msg,0,{get_fsm,1},{kv_vnode,1,1},{get,1}}},
 {deliver_msg,
     {msg,0,
         {kv_vnode,1,1},
         {get_fsm,1},
         {r,{ok,{r_object,<<"b">>,<<"k">>,
                    [{r_content,
                         {dict,0,16,16,8,80,48,
                             {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                             {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                               []}}},
                         <<0,0,0,3>>}],
                    [{<<0,0,0,2>>,{1,63472389426}}],
                    {dict,1,16,16,8,80,48,
                        {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                        {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                          [[clean|true]],
                          []}}},
                    undefined}},
            1,1}}},
 {deliver_msg,
     {msg,0,
         {get_fsm,1},
         {req,1},
         {ok,{r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,3>>}],
                 [{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {result,1,
     {req,0,1,{get,[{kv_vnode,1,1}]}},
     {ok,{r_object,<<"b">>,<<"k">>,
             [{r_content,
                  {dict,0,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                  <<0,0,0,3>>}],
             [{<<0,0,0,2>>,{1,63472389426}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}}},
 {deliver_msg,
     {msg,1,
         {req,4},
         {put_fsm,4},
         {put,
             [{kv_vnode,1,1}],
             {r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,4>>}],
                 [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {deliver_msg,
     {msg,0,
         {put_fsm,4},
         {kv_vnode,1,1},
         {put,
             {r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,4>>}],
                 [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined},
             4,4}}},
 {deliver_msg,{msg,0,{kv_vnode,1,1},{put_fsm,4},{dw,1,4}}},
 {deliver_msg,{msg,0,{kv_vnode,1,1},{put_fsm,4},{w,1,4}}},
 {deliver_msg,{msg,1,{put_fsm,4},{req,4},ok}},
 {result,1,{req,0,4,{put,[{kv_vnode,1,1}]}},ok},
 {deliver_msg,{msg,1,{kv_vnode,1,1},{put_fsm,3},{dw,1,3}}},
 {deliver_msg,{msg,0,{put_fsm,3},{req,3},ok}},
 {result,2,{req,0,3,{put,[{kv_vnode,1,1}]}},ok},
 {deliver_msg,{msg,0,{req,5},{get_fsm,5},{get,[{kv_vnode,1,1}]}}},
 {deliver_msg,{msg,0,{get_fsm,5},{kv_vnode,1,1},{get,5}}},
 {deliver_msg,
     {msg,1,
         {kv_vnode,1,1},
         {get_fsm,5},
         {r,{ok,{r_object,<<"b">>,<<"k">>,
                    [{r_content,
                         {dict,0,16,16,8,80,48,
                             {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                             {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                               []}}},
                         <<0,0,0,4>>}],
                    [{<<0,0,0,1>>,{1,63472389426}},
                     {<<0,0,0,2>>,{1,63472389426}}],
                    {dict,1,16,16,8,80,48,
                        {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                        {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                          [[clean|true]],
                          []}}},
                    undefined}},
            1,5}}},
 {deliver_msg,
     {msg,0,
         {get_fsm,5},
         {req,5},
         {ok,{r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,4>>}],
                 [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {result,1,
     {req,1,5,{get,[{kv_vnode,1,1}]}},
     {ok,{r_object,<<"b">>,<<"k">>,
             [{r_content,
                  {dict,0,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                  <<0,0,0,4>>}],
             [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}}},
 {deliver_msg,{msg,0,{req,6},{get_fsm,6},{get,[{kv_vnode,1,1}]}}},
 {deliver_msg,{msg,0,{get_fsm,6},{kv_vnode,1,1},{get,6}}},
 {deliver_msg,
     {msg,1,
         {kv_vnode,1,1},
         {get_fsm,6},
         {r,{ok,{r_object,<<"b">>,<<"k">>,
                    [{r_content,
                         {dict,0,16,16,8,80,48,
                             {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                             {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                               []}}},
                         <<0,0,0,4>>}],
                    [{<<0,0,0,1>>,{1,63472389426}},
                     {<<0,0,0,2>>,{1,63472389426}}],
                    {dict,1,16,16,8,80,48,
                        {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                        {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                          [[clean|true]],
                          []}}},
                    undefined}},
            1,6}}},
 {deliver_msg,
     {msg,0,
         {get_fsm,6},
         {req,6},
         {ok,{r_object,<<"b">>,<<"k">>,
                 [{r_content,
                      {dict,0,16,16,8,80,48,
                          {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                          {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                      <<0,0,0,4>>}],
                 [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
                 {dict,1,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                       [[clean|true]],
                       []}}},
                 undefined}}}},
 {result,1000001,
     {req,1000002,6,{get,[{kv_vnode,1,1}]}},
     {ok,{r_object,<<"b">>,<<"k">>,
             [{r_content,
                  {dict,0,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
                  <<0,0,0,4>>}],
             [{<<0,0,0,1>>,{1,63472389426}},{<<0,0,0,2>>,{1,63472389426}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}}}]
must: failed
[<<0,0,0,3>>] /= []
false

Commit:
Fixed model for counterexample - needed to add checks around out-of-order put results.

XXX Cid 2 GOT VALUES: []
XXX Cid 1 GOT VALUES: [<<0,0,0,3>>]
XXX Cid 1 PUT <<0,0,0,4>> OVER VALUES: [<<0,0,0,3>>]
XXX Cid 2 PUT <<0,0,0,3>> OVER VALUES: []
XXX Cid 1 GOT VALUES: [<<0,0,0,4>>]

The second put now checks the value is no longer in may.

Commit:
Another fix for out of order put responses.

Counterexample

[{[0,0,2,0,2,1,3],
  [{client,undefined,[get],undefined},
   {client,undefined,[update,get,update],undefined},
   {client,undefined,[get],undefined},
   {client,undefined,[update,get],undefined}],
  {params,undefined,0,2,2,undefined}}]

6> eqc:check(ec_eqc:prop(), eqc:current_counterexample()).
XXX Cid 2 GOT VALUES: []
XXX Cid 4 GOT VALUES: [<<0,0,0,4>>]
XXX Cid 2 PUT <<0,0,0,4>> OVER VALUES: []  MUST: [<<0,0,0,4>>] MAY: []
XXX Cid 2 GOT VALUES: [<<0,0,0,5>>]
XXX Cid 2 GOT VALUES: [<<0,0,0,5>>]
XXX Cid 2 PUT <<0,0,0,8>> OVER VALUES: [<<0,0,0,5>>]  MUST: [<<0,0,0,8>>,
                                                             <<0,0,0,4>>] MAY: [<<0,0,0,5>>]
XXX Cid 1 GOT VALUES: []
XXX Cid 4 PUT <<0,0,0,5>> OVER VALUES: [<<0,0,0,4>>]  MUST: [<<0,0,0,8>>] MAY: [<<0,0,0,5>>]
XXX Cid 4 GOT VALUES: [<<0,0,0,8>>]
XXX Cid 3 GOT VALUES: [<<0,0,0,8>>]
XXX Final GOT VALUES: [<<0,0,0,8>>] Must: [<<0,0,0,8>>] May: [<<0,0,0,5>>]
OK, passed the test.

Commit:
EC test passes for preflist made of primary nodes.

Commit:
Added handoff capability/final handoff but still using primaries in preflist.

Commit:
Added fallback nodes to preflists - get expected failures.

C1/C2 both update the value once on the primaries and C2 also updates
on fallback nodes only.  Because the primary VC is [{C1,1},{C2,1}] the
object on the fallback nodes is ignored with VC [{C2,1}]

61> ec_eqc:check().
XXXCid 1 GOT VALUES: []
XXXCid 1 PUT <<0,0,0,2>> OVER VALUES: []  MUST: [<<0,0,0,2>>] MAY: []
XXXCid 2 GOT VALUES: [<<0,0,0,2>>]
XXXCid 2 PUT <<0,0,0,4>> OVER VALUES: [<<0,0,0,2>>]  MUST: [<<0,0,0,4>>] MAY: [<<0,0,0,2>>]
XXXCid 2 GOT VALUES: []
XXXCid 2 PUT <<0,0,0,6>> OVER VALUES: []  MUST: [<<0,0,0,6>>,<<0,0,0,4>>] MAY: [<<0,0,0,2>>]
XXXFinal GOT VALUES: [<<0,0,0,4>>] Must: [<<0,0,0,6>>,<<0,0,0,4>>] May: [<<0,0,0,2>>]
Failed!
{[0],
 [{client,undefined,[{update,[0],[0]}],undefined},
  {client,undefined,[{update,[0],[0]},{update,[5],[0]}],undefined}],
 {params,undefined,0,1,2,undefined}}
Params:
{params,[{m,2},{n,2},{r,1},{w,2},{dw,2}]}
Clients:
[{client,1,
         [{req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined},
 {client,2,
         [{req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{get,[{kv_vnode,1,2},{kv_vnode,2,1}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined},
 {client,1000000,
         [{req,1000001,undefined,handoff_fallbacks},
          {req,1000002,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined}]

Commit:
Changed to only allowing one fallback per preflist, still has problems.

70> ec_eqc:check().
XXXCid 1 GOT VALUES: []
XXXCid 1 PUT <<0,0,0,2>> OVER VALUES: []  MUST: [<<0,0,0,2>>] MAY: []
XXXCid 2 GOT VALUES: []
XXXCid 2 PUT <<0,0,0,4>> OVER VALUES: []  MUST: [<<0,0,0,4>>,<<0,0,0,2>>] MAY: []
XXXCid 2 GOT VALUES: [<<0,0,0,4>>]
XXXCid 2 PUT <<0,0,0,6>> OVER VALUES: [<<0,0,0,4>>]  MUST: [<<0,0,0,6>>,
                                                            <<0,0,0,2>>] MAY: [<<0,0,0,4>>]
XXXCid 2 GOT VALUES: [<<0,0,0,6>>]
XXXCid 2 PUT <<0,0,0,8>> OVER VALUES: [<<0,0,0,6>>]  MUST: [<<0,0,0,8>>,
                                                            <<0,0,0,2>>] MAY: [<<0,0,0,4>>,
                                                                               <<0,0,0,6>>]
XXXCid 2 GOT VALUES: [<<0,0,0,6>>]
XXXCid 2 PUT <<0,0,0,10>> OVER VALUES: [<<0,0,0,6>>]  MUST: [<<0,0,0,10>>,
                                                             <<0,0,0,2>>,
                                                             <<0,0,0,8>>] MAY: [<<0,0,0,4>>,
                                                                                <<0,0,0,6>>]
XXXFinal GOT VALUES: [<<0,0,0,2>>,<<0,0,0,8>>] Must: [<<0,0,0,10>>,
                                                      <<0,0,0,2>>,
                                                      <<0,0,0,8>>] May: [<<0,0,0,4>>,
                                                                         <<0,0,0,6>>]
Failed!
{[0],
 [{client,undefined,[{update,[0],[2,5]}],undefined},
  {client,undefined,
          [{update,[0],[0]},
           {update,[0],[0]},
           {update,[0],[2,11]},
           {update,[0],[0]}],
          undefined}],
 {params,undefined,0,1,2,undefined}}
Params:
{params,[{m,2},{n,2},{r,1},{w,2},{dw,2}]}
Clients:
[{client,1,
         [{req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,2},{kv_vnode,2,2}]}}],
         undefined},
 {client,2,
         [{req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,2},{kv_vnode,2,2}]}},
          {req,0,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}},
          {req,0,undefined,{put,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined},
 {client,1000000,
         [{req,1000001,undefined,handoff_fallbacks},
          {req,1000002,undefined,{get,[{kv_vnode,1,1},{kv_vnode,2,2}]}}],
         undefined}]

Commit:
Even a single fallback causes problems.

114> ec_eqc:check().
XXXCid 1 GOT VALUES: [] FALLBACKS: []
XXXCid 1 PUT <<"C1-U1">> OVER [] MUST: [<<"C1-U1">>] MAY: [] FALLBACKS: []
XXXCid 1 GOT VALUES: [] FALLBACKS: [{1,2}]
XXXCid 1 PUT <<"C1-U2">> OVER [] MUST: [<<"C1-U2">>,<<"C1-U1">>] MAY: [] FALLBACKS: [{1,
                                                                                      2}]
XXXCid 1 GOT VALUES: [<<"C1-U1">>] FALLBACKS: []
XXXCid 1 PUT <<"C1-U3">> OVER [<<"C1-U1">>] MUST: [<<"C1-U3">>,<<"C1-U2">>] MAY: [<<"C1-U1">>] FALLBACKS: []
XXXFinal GOT VALUES: [<<"C1-U3">>] Must: [<<"C1-U3">>,<<"C1-U2">>] May: [<<"C1-U1">>] Fallbacks: []
Failed!
{[0],
 [{client,undefined,
          [{update,[0],[0]},{update,[6,1],[6,1]},{update,[0],[0]}],
          undefined}],
 {params,undefined,0,1,3,undefined}}
Params:
{params,[{m,3},{n,3},{r,1},{w,3},{dw,3}]}
Clients:
[{client,1,
         [{req,0,undefined,
               {get,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}]}},
          {req,0,undefined,
               {put,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}],
                    <<"C1-U1">>}},
          {req,0,undefined,
               {get,[{kv_vnode,1,2},{kv_vnode,2,2},{kv_vnode,3,3}]}},
          {req,0,undefined,
               {put,[{kv_vnode,1,2},{kv_vnode,2,2},{kv_vnode,3,3}],
                    <<"C1-U2">>}},
          {req,0,undefined,
               {get,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}]}},
          {req,0,undefined,
               {put,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}],
                    <<"C1-U3">>}}],
         undefined},
 {client,1000000,
         [{req,1000001,undefined,handoff_fallbacks},
          {req,1000002,undefined,
               {get,[{kv_vnode,1,1},{kv_vnode,2,2},{kv_vnode,3,3}]}}],
         undefined}]

Commit:
Tried generating fail message on failed-due-to-out-of-date-vclock writes

Reverted to generating preference lists using all nodes.
Still fails.

149> ec_eqc:check().
XXXCid 1 GOT VALUES: [] FALLBACKS: []
XXXCid 1 PUT <<"C1-U1">> OVER [] MUST: [] MAY: [] FALLBACKS: [{1,2},
                                                              {2,3},
                                                              {3,1}]
XXXCid 2 GOT VALUES: [] FALLBACKS: []
XXXCid 2 PUT <<"C2-U1">> OVER [] MUST: [] MAY: [] FALLBACKS: [{1,2},
                                                              {2,3},
                                                              {3,1}]
XXXCid 2 GOT VALUES: [] FALLBACKS: []
XXXCid 2 PUT <<"C2-U2">> OVER [] MUST: [<<"C2-U2">>] MAY: [] FALLBACKS: []
XXXFinal GOT VALUES: [<<"C1-U1">>,<<"C2-U1">>] Must: [<<"C2-U2">>] May: [] Fallbacks: []
Failed!
{[0],
 [{client,undefined,[{update,[0],[16]}],undefined},
  {client,undefined,[{update,[0],[1]},{update,[0],[0]}],undefined}],
 {params,undefined,0,1,2,undefined}}
Params:
{params,[{m,3},{n,3},{r,2},{w,2},{dw,2}]}

Commit:
Still getting inconsistencies despite writing to W=all, DW=all when there are fallbacks.

Commit:
Changed put FSM to do dynamo-style coordinator vnode put first then remote.

Commit:
Reverted to generating parameters again - put still has W/DW=all

Commit:
Make w/dw generated as part of the test.

Commit:
Added read repair.

Commit:
Got handoff working with object going immediately undefined.

In the real code the handoff transmission/deletion is async.
Handoff starts, network connection gets set up, a fold kicks off
and once all data is transmitted the vnode does the drop.  It could
service other requests in the mean time.  In fact it looks
like the data is dropped on handoff completion, even if something
has arrived in the mean time.

Commit:
Made handoff/cleanup async.

Commit:
Changed coordinating vnode to increment clock if necessary.

Commit:
Changed put FSM / kv vnode to support dynamo style puts.

Put FSM test currently broken.

Commit:
Fixed missing state update and incorrectly accessing vector clock counters.

Commit:
Updated put FSM eqc tests for vclock changes.

beerriot commented on src/riak_kv_put_fsm.erl:
```diff
@@ -244,35 +267,81 @@ precommit(timeout, State = #state{precommit = [Hook | Rest], robj = RObj}) ->
+    case riak_kv_put_core:enough(UpdPutCore1) of
+        true ->
+            {Reply, UpdPutCore2} = riak_kv_put_core:response(UpdPutCore1),
+            process_reply(Reply, StateData#state{putcore = UpdPutCore2});
```
This path short-circuits to either `postcommit` or `finish`.  Since it never goes through `execute_remote`, the replicas are never written.

Test code (on an N=3 bucket):

```
{B, K} = {<<"az303">>, <<"w=1">>},
{ok, C} = riak:local_client(),
ok = C:put(riak_object:new(B, K, ok), 1).
… wait as long as you like …
C:get(B, K, 3).  % R=2 shows the same behavior
```

Expected: `{ok, Object}`

Actual: `{error, notfound}`

jonmeredith replied:
Nice catch, thanks!

beerriot commented on src/riak_kv_vnode.erl:
```diff
@@ -282,6 +306,52 @@ handle_exit(_Pid, _Reason, State) ->
+    UpdVC = riak_object:vclock(UpdObj),
+
+    %% Valid coord put replacing current object
+    case get_counter(CoordId, UpdVC) > get_counter(CoordId, CurVC) andalso
```
Something that has bugged me is that puts are not truly idempotent.  That is, `C:put(O1), C:put(O1)` creates an object with two sibling values that happen to be equal[1], instead of ignoring the second put.  It seems like a vector clock comparison at the correct time should catch this.  It didn't before this patch because the vclock was incremented immediately, so the comparison found equal vclocks, not sequential ones.  I'm not sure it can again after this patch, because two writes from two clients may serialize on the same vnode.  But, I thought I'd bring up the topic in case it generated more useful discussion in the context of this patch.

[1] The objects are only logically equal, and not `=:=` because of Last-Modified metadata.

slfritchie replied:
After this patch:

```
% ./dev/dev1/bin/riak attach
Attaching to /tmp//Users/fritchie/b/src/riak/dev/dev1/erlang.pipe.1 (^D to exit)

nodes().
['dev3@127.0.0.1','dev2@127.0.0.1']
(dev1@127.0.0.1)2> O1 = riak_object:new(<<"b">>, <<"k">>, <<"val master">>).
{r_object,<<"b">>,<<"k">>,
          [{r_content,{dict,0,16,16,8,80,48,
                            {[],[],[],[],[],[],[],[],[],[],[],[],[],[],...},
                            {{[],[],[],[],[],[],[],[],[],[],[],[],...}}},
                      <<"val master">>}],
          [],
          {dict,1,16,16,8,80,48,
                {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],...},
                {{[],[],[],[],[],[],[],[],[],[],[],[],[],...}}},
          undefined}
(dev1@127.0.0.1)3> {ok, C1} = riak:local_client().
{ok,{riak_client,'dev1@127.0.0.1',<<5,231,185,7>>}}
(dev1@127.0.0.1)4> C1:put(O1).
ok
(dev1@127.0.0.1)5> C1:put(O1).
ok
(dev1@127.0.0.1)6> C1:get(<<"b">>, <<"k">>).
{ok,{r_object,<<"b">>,<<"k">>,
              [{r_content,{dict,2,16,16,8,80,48,
                                {[],[],[],[],[],[],[],[],[],[],[],[],...},
                                {{[],[],[],[],[],[],[],[],[],[],...}}},
                          <<"val master">>}],
              [{'dev1@127.0.0.1',{2,63477892858}}],
              {dict,1,16,16,8,80,48,
                    {[],[],[],[],[],[],[],[],[],[],[],[],[],...},
                    {{[],[],[],[],[],[],[],[],[],[],[],...}}},
              undefined}}
(dev1@127.0.0.1)7> [Quit]
```

beerriot commented on test/ec_eqc.erl:
```diff
@@ -0,0 +1,844 @@
+                    riak_object:merge(CurObj, UpdObj)
+            end
+    end.
+
```
I think the code below is duplicated from riak_kv_vnode.  Can the function just be exported from there instead?

beerriot commented on test/put_fsm_eqc.erl:
```diff
@@ -183,7 +183,7 @@ vputsecond() ->
     ?SHRINK(frequency([{16, primary},
                        {4, fallback},
-                       {1, down}]),
+                       {0, down}]),
```
What's the effect of changing this `{1, down}` to `{0, down}`.  In conversation you mentioned something about the test no longer testing down nodes, and that the line should have been completely scrubbed.  True?

beerriot commented:
I'm a little worried about whether `node()` is unique enough for actor ID in the vclock.  The scenarios I'm thinking of are moments when the chosen coordinator may change.

The first is during handoff caused by cluster membership change.  Imagine that at T0, node N1 owns partition P.  N1 coordinates a few writes for key K, so key K has vector clock, [{N1, 5}].  At T1, we add node N2, and partition P shifts to node N2.  When handoff is complete, node N1 drops partition P.  At T2, we change cluster membership again in some way that causes partition P to shift back to node N1.  Partition P on node N1 may be asked to coordinate writes before handoff completes.  If this happens, a vclock for key K may be created such that it is an ancestor of the K's vclock stored on node N2.  When handoff completes, data stored under the ancestor vclock will be lost.  Replica partitions will also ignore the older vclock.  This would not happen for any write that was preceded by a successful read (and therefore included the existing vclock), but would happen for reads that failed as "not found", as well as for misbehaved clients.

The other case I'm thinking of is unbalanced rings, where one node owns more than one partition in a preflist.  In that state, only one of those partitions will be used as coordinator (because of the "head" match in riak_kv_put_fsm), but if that coordinator partition shifts to a different node, then the extra partition will become the coordinator.  If it has missed any updates, it may begin generating old vclocks (again, with the exception of successful reads-before-writes).  This same problem might happen if a node loses one partition in a given preflist, but gains another.

I've not written failing test code for either of these situations, so I might just be flat out wrong.  They're also a little corner-case-y, and don't apply to well-behaved read-then-write operations.  But, I still wonder if it wouldn't be better to generate a unique actor ID at vnode startup, and to use that instead of `node()`.  Even something as simple as a combination of node+partition+start_time could work.  Vector clocks would grow a little more than when using just node, but it would still be bounded by n_val \* membership_changes, if done smartly, or even n_val \* node_restarts if done naively.

beerriot commented on src/riak_object.erl:
```diff
@@ -445,8 +445,8 @@ syntactic_merge(CurrentObject, NewObject, FromClientId, Timestamp) ->
                 false ->
-                    increment_vclock(
-                      merge(CurrentObject, NewObject), FromClientId, Timestamp)
+                    %%increment_vclock(
```
This change makes `riak_object:merge2_test` fail (it expects a new client id to appear in the vclock actor list after merging).

beerriot commented:
As far as client-visible changes go, this patch makes ClientId in riak_client (and therefore its HTTP and PB equivalents) obsolete and unused.  Everything should continue to work as expected, but it may be worth pinging the community to warn people that are doing their own vclock introspection.

slfritchie commented:
Jon: I'm seeing two EUnit test failures:

```
  riak_object: merge2_test...*failed*
::error:{badmatch,[node1,node2]}
  in function riak_object:merge2_test/0
```

And the second is:

```
module 'put_fsm_eqc'
  put_fsm_eqc:71: eqc_test_...ok
  put_fsm_eqc:72: eqc_test_...[0.001 s] ok
  put_fsm_eqc:73: eqc_test_...[0.009 s] ok
  put_fsm_eqc:74: eqc_test_...ok
  put_fsm_eqc:76: eqc_test_...[0.002 s] ok
  put_fsm_eqc:83: eqc_test_...FSM died:
{function_clause,
    [{gen_server,cast,
         [{riak_kv_vnode_master,<0.4073.0>},
[...]
Shrinking...(9 times)
{quorum,default,one,
    [{ancestor,
         {r_object,<<"x">>,<<"4">>,
             [{r_content,
                  {dict,1,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],
                        [[<<"X-Riak-Last-Modified">>|{1310,672109,470130}]],
                        [],[]}}},
                  <<"ancestor">>}],
             [],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}},
     {brother,
         {r_object,<<"x">>,<<"4">>,
             [{r_content,
                  {dict,1,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],
                        [[<<"X-Riak-Last-Modified">>|{1310,672109,470142}]],
                        [],[]}}},
                  <<"brother">>}],
             [{<<"bro!">>,{1,63477891309}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}},
     {sister,
         {r_object,<<"x">>,<<"4">>,
             [{r_content,
                  {dict,1,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],
                        [[<<"X-Riak-Last-Modified">>|{1310,672109,470148}]],
                        [],[]}}},
                  <<"sister">>}],
             [{<<"sis!">>,{1,63477891309}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}},
     {otherbrother,
         {r_object,<<"x">>,<<"4">>,
             [{r_content,
                  {dict,1,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],
                        [[<<"X-Riak-Last-Modified">>|{1310,672109,470157}]],
                        [],[]}}},
                  <<"otherbrother">>}],
             [{<<"bro2">>,{1,63477891309}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}},
     {current,
         {r_object,<<"x">>,<<"4">>,
             [{r_content,
                  {dict,1,16,16,8,80,48,
                      {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                      {{[],[],[],[],[],[],[],[],[],[],[],[],[],
                        [[<<"X-Riak-Last-Modified">>|{1310,672109,470163}]],
                        [],[]}}},
                  <<"current">>}],
             [{<<"bro!">>,{1,63477891309}},
              {<<"bro2">>,{1,63477891309}},
              {<<"sis!">>,{1,63477891309}}],
             {dict,1,16,16,8,80,48,
                 {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                 {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                   [[clean|true]],
                   []}}},
             undefined}}],
    0,
    [{notfound,w,0,dw,0,primary}],
    [],false,[],[]}
BucketProps: [{n_val,1},
              {allow_mult,false},
              {chash_keyfun,{riak_core_util,chash_std_keyfun}},
              {pw,0},
              {w,quorum},
              {dw,quorum}]
VPutReplies = [{1,w},{1,{dw,notfound,notfound}}]
PrefList2: [{{1,<0.4073.0>},primary}]
N: 1 PW: quorum RealPW: 1 W:default RealW: 1 DW: one RealDW: 1 EffDW: 1 Pid: <0.4164.0>
Options: [{timeout,200},{dw,one},{w,default},{pw,quorum}]
Precommit: []
Postcommit: []
Object: {r_object,<<"x">>,<<"4">>,
                  [{r_content,{dict,1,16,16,8,80,48,
                                    {[],[],[],[],[],[],[],[],[],[],[],[],[],
                                     [],[],[]},
                                    {{[],[],[],[],[],[],[],[],[],[],[],[],[],
                                      [[<<"X-Riak-Last-Modified">>|
                                        {1310,672109,470163}]],
                                      [],[]}}},
                              <<"current">>}],
                  [{<<"bro!">>,{1,63477891309}},
                   {<<"bro2">>,{1,63477891309}},
                   {<<"sis!">>,{1,63477891309}}],
                  {dict,1,16,16,8,80,48,
                        {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                        {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                          [[clean|true]],
                          []}}},
                  <<"current">>}
Expected Object: noreply
History: []
Expected: {error,timeout} Res: timeout
PostCommits:
Expected: []
Got: []
result: failed
timeout /= {error,timeout}
*failed*
::error:{assertEqual_failed,
          [{module,put_fsm_eqc},
           {line,83},
           {expression,
               "quickcheck ( numtests ( 1000 , ? QC_OUT ( prop_basic_put ( ) ) ) )"},
           {expected,true},
           {value,false}]}
  in function put_fsm_eqc:'-eqc_test_/0-fun-13-'/1
```

jonmeredith commented:
Hey Scott,

Pull the az303 branch for riak core to fix the put FSM test - I may need to look at the merge test.

Jon

slfritchie commented:
AZ303, right, nothing like reading the introduction.  Yup, that fixed it.  And the other
failing eunit test case is something that one of Bryan's comments mentions also.

The complexity of testing that's in ec_eqc.erl is ... big.  The must vs. have stuff is
necessary, but hopefully there's a better way than reimplementing a fraction of the
Erlang VM also.  Bah.  I'm thinking that's a good dragon to try to slay.

Commit:
Updated merge2_test for changes to riak_object:syntactic_merge().

Commit:
Changed ec_eqc to handoff to random partition, not just owner.

Used to demonstrate issue spotted by Bryan,
https://github.com/basho/riak_kv/pull/127#issuecomment-1554900
This is equivalent to the ownership change described in the comment
and lets us test sequences like this:

dev2: create object
dev2: update object (object now [{dev2,2}],v1b
dev2: reassign partition
dev2: wait for transfer to complete
dev1: get object to make sure it's there
dev1: shutdown
dev2: get object, check notfound
dev2: create object (vclock now [{dev2,1],v1)
dev2: check object
dev1: startup, wait for handoff to complete.
dev1: sure enough, v1b is the object that is found.

But without introducing owners to the model.  I don't think it's necessary
as the final handoff provides the synchronization point needed to check
eventual consistency.  Before that point any vnode should handoff to any
other.

Commit:
Stopped ec_eqc test from deleting primary partitions after transfer.

Commit:
Prepared ec_eqc datastructures for modelling ownership.

Commit:
ec_eqc test now demonstrates issues around dropping vnodes.

Commit:
Updated ec_eqc and put fsm/vnode to add vnode epoch to vclock node id.

Fixes the case identified by Bryan in the pull request comments, where
a vnode can transfer ownership (deleting the local copy of the data) then
be used as a partitioned fallback.  Previously that would have re-used
a vclock entry, but now the vnode epoch is included in the clock
to guarantee uniqueness.

Commit:
Cleaned up the code to create a per-node/vnode vclock id.

Also modified vclock:get_counter to behave the way the wrapper
was working - no other consumers were using it.

Commit:
Updated put FSM eqc mock vnode tests to do match coord puts in kv_vnode.

Commit:
Extended put FSM eqc test to ensure correct number of put requests received.

Commit:
Fixed case where put requests not sent to vnodes W/DW=1.

Required changes to the put EQC model to work out how many
responses should be expected and added a special case for
local vnode failure.

Commit:
Added check for riak_kv:vnode_vclocks app var to enable new behavior.

When running with it set to anything but true it will increment
by the client id before doing the local put - this is only really
needed for forwarding to other nodes if not in the preference list
as the local node will support a vnode coord put.

jonmeredith commented:
Replaced by https://github.com/basho/riak_kv/pull/157 - rebased version against master.
Did not want to risk losing detailed review comments by rebase/pushing
